[RPC Framework] Create a separate remote module template when moving CPU tensors to a cuda device is not enabled

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57414 [RPC Framework] Clang-format remote_module.py and instantiator.py
* **#57413 [RPC Framework] Create a separate remote module template when moving CPU tensors to a cuda device is not enabled**

An internal test fails because somehow `Tuple[()]` is not considered compatible with `Tuple[Any]` in TorchScript, even if the code that involves this type of variables is not executed at all.

Therefore, create separate templates for instantiation to avoid typing check failure. This can address the FIXME left in https://github.com/pytorch/pytorch/pull/57288

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28138864](https://our.internmc.facebook.com/intern/diff/D28138864/)

Commit:
[RPC Framework] Create a separate remote module template when moving CPU tensors to a cuda device is not enabled

An internal test fails because somehow `Tuple[()]` is not considered compatible with `Tuple[Any]` in TorchScript, even if the code that involves this type of variables is not executed at all.

Therefore, create separate templates for instantiation to avoid typing check failure. This can address the FIXME left in https://github.com/pytorch/pytorch/pull/57288

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28138864](https://our.internmc.facebook.com/intern/diff/D28138864/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Create a separate remote module template when moving CPU tensors to a cuda device is not enabled"


An internal test fails because somehow `Tuple[()]` is not considered compatible with `Tuple[Any]` in TorchScript, even if the code that involves this type of variables is not executed at all.

Therefore, create separate templates for instantiation to avoid typing check failure. This can address the FIXME left in https://github.com/pytorch/pytorch/pull/57288

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28138864](https://our.internmc.facebook.com/intern/diff/D28138864/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Create a separate remote module template when moving CPU tensors to a cuda device is not enabled"


An internal test fails because somehow `Tuple[()]` is not considered compatible with `Tuple[Any]` in TorchScript, even if the code that involves this type of variables is not executed at all.

Therefore, create separate templates for instantiation to avoid typing check failure. This can address the FIXME left in https://github.com/pytorch/pytorch/pull/57288

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28138864](https://our.internmc.facebook.com/intern/diff/D28138864/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Create a separate remote module template when moving CPU tensors to a cuda device is not enabled"


An internal test fails because somehow `Tuple[()]` is not considered compatible with `Tuple[Any]` in TorchScript, even if the code that involves this type of variables is not executed at all.

Therefore, create separate templates for instantiation to avoid typing check failure. This can address the FIXME left in https://github.com/pytorch/pytorch/pull/57288

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28138864](https://our.internmc.facebook.com/intern/diff/D28138864/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Create a separate remote module template when moving CPU tensors to a cuda device is not enabled"


An internal test fails because somehow `Tuple[()]` is not considered compatible with `Tuple[Any]` in TorchScript, even if the code that involves this type of variables is not executed at all.

Therefore, create separate templates for instantiation to avoid typing check failure. This can address the FIXME left in https://github.com/pytorch/pytorch/pull/57288

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28138864](https://our.internmc.facebook.com/intern/diff/D28138864/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/nn/jit/templates/remote_module_template.py:
```diff
@@ -1,19 +1,63 @@
 {jit_script_decorator}
+"""
+
+# This template may cause typing error (the mismatch between ``Tuple[()]`` and ``Tuple[Any]``)
```
can you add a TODO here to make sure we merge these two template together in the future when TorchScript syntax improves?

wanchaol reviewed (approved):
looks good to me

Commit:
Update on "[RPC Framework] Create a separate remote module template when moving CPU tensors to a cuda device is not enabled"


An internal test fails because somehow `Tuple[()]` is not considered compatible with `Tuple[Any]` in TorchScript, even if the code that involves this type of variables is not executed at all.

Therefore, create separate templates for instantiation to avoid typing check failure. This can address the FIXME left in https://github.com/pytorch/pytorch/pull/57288

#Closes: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28138864](https://our.internmc.facebook.com/intern/diff/D28138864/)

[ghstack-poisoned]
